pkg/ottl/parser.go

```diff
@@ -56,15 +56,53 @@ func (s *Statement[K]) Execute(ctx context.Context, tCtx K) (any, bool, error) {
 	return result, condition, nil
 }
 
-func NewParser[K any](functions map[string]interface{}, pathParser PathExpressionParser[K], enumParser EnumParser, telemetrySettings component.TelemetrySettings) Parser[K] {
+func NewParser[K any](
+	functions map[string]interface{},
+	pathParser PathExpressionParser[K],
+	options ...Option,
+) Parser[K] {
+	var cfg config
+	for _, opt := range options {
+		cfg = opt.apply(cfg)
+	}
+
 	return Parser[K]{
 		functions:         functions,
 		pathParser:        pathParser,
-		enumParser:        enumParser,
-		telemetrySettings: telemetrySettings,
+		enumParser:        cfg.enumParser,
+		telemetrySettings: cfg.telemetrySettings,
 	}
 }
 
+type config struct {
+	enumParser        EnumParser
+	telemetrySettings component.TelemetrySettings
+}
+
+type Option interface {
+	apply(config) config
+}
+
+type parserOptionFunc func(config) config
```

Comment:
I think it is cleaner if `parserOptionFunc` takes a pointer to a Parser and returns an error.  This way the option function can do the modification of parser object itself we we don't have to have a new `config` struct.

Reply:
I tried to use a pointer to Parser, but since it has a type parameter it was not that easy to make it work, thus, I had to introduce this indirection with `config`.
```
func NewParser[K any](
	functions map[string]interface{},
	pathParser PathExpressionParser[K],
	options ...Option,
) Parser[K] {
	p := Parser[K]{
		functions:  functions,
		pathParser: pathParser,
	}
	for _, opt := range options {
		opt(&p) <---------- opt expects *Parser[any]
	}
	return p
}

type Option func(*Parser[any])

func WithEnumParser(parser EnumParser) Option {
	return func(p *Parser[any]) {
		p.enumParser = parser
	}
}
```

Since `Parser[K]` has a type parameter then

Reply:
Can `Option` and `With` be updated to use generics also?  I think something like

```go
func NewParser[K any](
	functions map[string]interface{},
	pathParser PathExpressionParser[K],
	options ...Option[K],
) Parser[K] {
	p := Parser[K]{
		functions:  functions,
		pathParser: pathParser,
	}
	for _, opt := range options {
		opt(&p)
	}
	return p
}

type Option[K any] func(*Parser[K])

func WithEnumParser[K any](parser EnumParser) Option[K] {
	return func(p *Parser[K]) {
		p.enumParser = parser
	}
}
```
will work.

Reply:
Yes, that would work, but on an invocation side one would need to provide a type parameter for `With...` functions even though it will be irrelevant for most cases. E.g. 

```
return ottl.NewParser[TransformContext](
	functions,
	parsePath,
	ottl.WithEnumParser[TransformContext](parseEnum),
	ottl.WithTelemetrySettings[TransformContext](telemetrySettings),
)
```

Reply:
Ya that is pretty annoying :/.  For users of the ottl context's it won't be bad since the generics are abstracted away by each context's NewParser function at least. @evan-bradley @kentquirk @bogdandrutu what are your opinions?

@kovrus for the `config` solution, is it possible to do

```go
func NewParser[K any](
	functions map[string]interface{},
	pathParser PathExpressionParser[K],
	options ...Option,
) Parser[K] {
	var cfg config
	for _, opt := range options {
		opt(&cfg)
	}

	return Parser[K]{
		functions:         functions,
		pathParser:        pathParser,
		enumParser:        cfg.enumParser,
		telemetrySettings: cfg.telemetrySettings,
	}
}

type config struct {
	enumParser        EnumParser
	telemetrySettings component.TelemetrySettings
}

type Option func(*config)

func WithEnumParser(parser EnumParser) Option {
	return func(c *config) {
		c.enumParser = parser
	}
}

func WithTelemetrySettings(settings component.TelemetrySettings) Option {
	return func(c *config) {
		c.telemetrySettings = settings
	}
}
```

Reply:
I don't have a strong opinion either way, but I would lean toward having the `Option` type operate on the parser struct directly, even if it means we have to pass extra type arguments when creating a parser. I think the two main advantages are:

1. There's fewer changes to make when adding an option, since we can just work directly with the parser and don't have to manage an extra struct.
2. The public `Option` type doesn't refer to an internal type (`config`), or alternatively we don't have to expose a new type if we made `config` public.